### PR TITLE
Subtitle Styling

### PR DESCRIPTION
**Pull request by alr2413**
Add support for subtitle styling #190.
Also split the plugin swift code into multiple files to improve readability and management.
Fix a minor bug where buffering value is not set to false when the media is playing.

### CODE REVIEW

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart [hunk 1]
```diff
@@ -0,0 +1,90 @@
+class VlcSubtitleOptions {
+  final List<String> options;
+
+  VlcSubtitleOptions(this.options);
+
+  /// Set subtitle font (must be accessible by vlc)
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:6]: Perhaps we could add some documentation on how to make the font available to the native lib?

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    return '--freetype-opacity=$opacity';
+  }
+
+  /// Set subtitle color
+  static String color(int color) {
+    return '--freetype-color=$color';
+  }
+
+  /// Set subtitle text direction {0 (Left to right), 1 (Right to left), 2 (Auto)}
+  static String textDirection(int direction) {
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:37]: We could probably turn this into an enum.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Looking good! 
Just one small fix and a couple of suggestions.

Looks like we're not making the new options available in `flutter_vlc_player.dart` exports. I needed to add this to be able to use them:
image

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart (hunk 1 above)
Comment (alr2413) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:6]: we can but i really don't know how it works.
i tried to set the subtitle font in vlc but it throws the font not found.
the only options i tested are font size, color, thickness & opacity.

Comment (solid-yuriiprykhodko) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:6]: Yeah, this looks like a tricky one.

I've tried a variety of things:
1. Loading a font to filesystem and pointing the lib to its location.
2. Using a system font by its name (e.g. 'Roboto', 'AndroidClock').
3. Using a system font by its filesystem path ('`/system/fonts/AndroidClock.ttf`').
4. Loading a font via `google_fonts` package and using them by name.
5. Bundling fonts with and iOS app.

None of the approaches seemed to work. I'm not sure we want to keep these options exposed, unless we find a way to get them to actually work.

Comment (solid-yuriiprykhodko) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:6]: I've also tested the following:
1. Font size, color, opacity, boldness.
2. Background color, opacity.
3. Text direction.
4. Outline opacity, color, thickness.
5. Shadow color, distance and angle.

Looks like the library handles the out-of-range input clipping by itself, so nonsensical inputs behave as if they're clipped to the closest constraint ((opacity `-200.0` -> `0.0`)).

Comment (alr2413) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:10]: did you test that with an external .srt file?!

Comment (solid-yuriiprykhodko) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:10]: Yeah, I tested with an `.srt`. For built-in subs the font color does not change, but the outline color option works OK.

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 3]
```diff
@@ -92,6 +97,14 @@ class _SingleTabState extends State<SingleTab> {
             advanced: VlcAdvancedOptions([
               VlcAdvancedOptions.networkCaching(2000),
             ]),
+            subtitle: VlcSubtitleOptions([
+              VlcSubtitleOptions.boldStyle(true),
+              VlcSubtitleOptions.fontSize(30),
+              // works only on external subs
+              VlcSubtitleOptions.color(VlcSubtitleColor.Navy),
+              // works only on external subs
+              VlcSubtitleOptions.outlineColor(VlcSubtitleColor.Yellow),
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:106]: Outline color changes with built-in subs for me. What platform are you testing on?

File: flutter_vlc_player/pubspec.yaml [hunk 4]
```diff
@@ -22,7 +22,9 @@ dependencies:
   cryptoutils: ^0.4.0
   meta: ^1.2.4
   pedantic: ^1.9.0
-  flutter_vlc_player_platform_interface: ^1.0.3
+#  flutter_vlc_player_platform_interface: ^1.0.4
+  flutter_vlc_player_platform_interface:
+    path: ../flutter_vlc_player_platform_interface
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/pubspec.yaml:27]: Let's not forget to change that back before merging.
```suggestion
  flutter_vlc_player_platform_interface: ^1.0.4
```

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    return '--freetype-background-color=${color.value}';
+  }
+
+  /// Set subtitle outline opacity [0 .. 255]
+  static String outlineOpacity(int opacity) {
+    return '--freetype-outline-opacity=$opacity';
+  }
+
+  /// Set subtitle outline color
+  static String outlineColor(VlcSubtitleColor color) {
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:66]: I like that it's a bit more structured than just an `int` input, but the current implementation limits the user to a predefined set of colors.

We should probably provide a way to create custom colors. We could do that by turning a `VlcSubtitleColor` enum into an ADT, with a public constructor for custom colors, at the same time keeping the predefined values as `static const` fields on the class.
```dart
@immutable
class VlcSubtitleColor {
  final int value;

  static const navy = VlcSubtitleColor(12345);

  const VlcSubtitleColor(this.value);
  const VlcSubtitleColor.rgb({
    int red = 0,
    int green = 0,
    int blue = 0,
  }): value = ...;
}
```

File: flutter_vlc_player_platform_interface/lib/src/enums/subtitles/vlc_subtitle_text_direction.dart [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  AUTO,
+}
+
+extension VlcSubtitleTextDirectionExtensionMap on VlcSubtitleTextDirection {
+  static const valueMap = {
+    VlcSubtitleTextDirection.LTR: 0,
+    VlcSubtitleTextDirection.RTL: 1,
+    VlcSubtitleTextDirection.AUTO: 2,
+  };
+  int get value => valueMap[this];
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/enums/subtitles/vlc_subtitle_text_direction.dart:13]: Let's use Effective Dart naming convention for enums: https://dart-lang.github.io/linter/lints/constant_identifier_names.html
```suggestion
enum VlcSubtitleTextDirection {
  ltr,
  rtl,
  auto,
}

extension VlcSubtitleTextDirectionExtensionMap on VlcSubtitleTextDirection {
  static const valueMap = {
    VlcSubtitleTextDirection.ltr: 0,
    VlcSubtitleTextDirection.rtl: 1,
    VlcSubtitleTextDirection.auto: 2,
  };
  int get value => valueMap[this];
```

File: flutter_vlc_player_platform_interface/lib/src/enums/subtitles/vlc_subtitle_color.dart [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    VlcSubtitleColor.Fuchsia: 16711935,
+    VlcSubtitleColor.Yellow: 16776960,
+    VlcSubtitleColor.Olive: 8421376,
+    VlcSubtitleColor.Green: 32768,
+    VlcSubtitleColor.Teal: 32896,
+    VlcSubtitleColor.Lime: 65280,
+    VlcSubtitleColor.Purple: 8388736,
+    VlcSubtitleColor.Navy: 128,
+    VlcSubtitleColor.Blue: 255,
+    VlcSubtitleColor.Aqua: 65535,
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/enums/subtitles/vlc_subtitle_color.dart:37]: Let's use Effective Dart naming convention for enums: https://dart-lang.github.io/linter/lints/constant_identifier_names.html
```suggestion
enum VlcSubtitleColor {
  black,
  gray,
  silver,
  white,
  maroon,
  red,
  fuchsia,
  yellow,
  olive,
  green,
  teal,
  lime,
  purple,
  navy,
  blue,
  aqua,
}

extension VlcSubtitleColorExtensionMap on VlcSubtitleColor {
  static const valueMap = {
    VlcSubtitleColor.black: 0,
    VlcSubtitleColor.gray: 8421504,
    VlcSubtitleColor.silver: 12632256,
    VlcSubtitleColor.white: 16777215,
    VlcSubtitleColor.maroon: 8388608,
    VlcSubtitleColor.red: 16711680,
    VlcSubtitleColor.fuchsia: 16711935,
    VlcSubtitleColor.yellow: 16776960,
    VlcSubtitleColor.olive: 8421376,
    VlcSubtitleColor.green: 32768,
    VlcSubtitleColor.teal: 32896,
    VlcSubtitleColor.lime: 65280,
    VlcSubtitleColor.purple: 8388736,
    VlcSubtitleColor.navy: 128,
    VlcSubtitleColor.blue: 255,
    VlcSubtitleColor.aqua: 65535,
```

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  /// Set subtitle font size
+  static String fontSize(int size) {
+    return '--freetype-fontsize=$size';
+  }
+
+  /// Set subtitle rel font size
+  static String relFontSize(int size) {
+    return '--freetype-rel-fontsize=$size';
+  }
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:28]: Let's use full naming where we can. I've had to Google it to verify that it's a _relative_ font size setting :).
```suggestion
  /// Set subtitle relative font size
  static String relativeFontSize(int size) {
    return '--freetype-rel-fontsize=$size';
  }
```

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Nice improvements! Really like the more structured inputs. 
A couple of comments on how we could improve it even more.

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 9]
```diff
@@ -84,6 +84,11 @@ class _SingleTabState extends State<SingleTab> {
           hwAcc: HwAcc.FULL,
           onInit: () async {
             await _controller.startRendererScanning();
+            // just a sample subtitle file
+            await _controller.addSubtitleFromNetwork(
+              'https://gist.githubusercontent.com/alr2413/9bfc7f3439ea473cbdb3ccb76d7613a3/raw/a172c10aa902078eeda9a1345c79bad9ac367d3d/subtitle_sample.srt',
+              isSelected: true,
+            );
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:91]: Let's not forget to remove this, too. The subtitles don't match the sample video, which might be confusing for people that are just trying out the example app.
```suggestion
```

File: flutter_vlc_player/example/lib/single_tab.dart (hunk 3 above)
Comment (alr2413) [reply flutter_vlc_player/example/lib/single_tab.dart:106]: yea, you'r right.

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart (hunk 1 above)
Comment (alr2413) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:10]: yea, you're right.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Looks good! 
Just one more changes before we're good to merge.

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    return '--freetype-outline-opacity=$opacity';
+  }
+
+  /// Set subtitle outline color
+  static String outlineColor(VlcSubtitleColor color) {
+    return '--freetype-outline-color=${color.value}';
+  }
+
+  /// Set subtitle outline thickness {0 (None), 2 (Thin), 4 (Normal), 6 (Thick)}
+  static String outlineThickness(VlcSubtitleThickness thickness) {
```
[the remark this answers is not in the document]
Comment (alr2413) [reply flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_subtitle_options.dart:71]: fixed.

**Review by solid-yuriiprykhodko — APPROVED** [review]
Thanks, great work on this PR!

**Comment by demrebicer** [comment]
@alr2413 Can you check this issue ? (https://github.com/solid-software/flutter_vlc_player/issues/362)

Now this PR not working.

**Comment by alr2413** [comment]
> @alr2413 Can you check this issue ? (#362)
> 
> Now this PR not working.

Could you please check if it works with raw vlc options instead of using the provided ones?

**Comment by demrebicer** [comment]
I am really sorry. I don't know how to do this.

**Pull request merged**